core/src/proxy/auth.rs

```diff
@@ -42,206 +40,143 @@ impl Interceptor for AuthInterceptor {
-                .as_ref()
-                .map(|ts| ts.seconds)
-                .unwrap_or_default();
-
-            let now = Utc::now().timestamp();
-
-            let should_refresh_access = access_token_expiry.checked_sub(now).ok_or_else(|| {
-                ProxyError::InvalidData("Received invalid access_token expiration".to_string())
-            })? <= REFRESH_WITHIN_SECS;
-            let should_generate_new_tokens =
-                refresh_token_expiry.checked_sub(now).ok_or_else(|| {
-                    ProxyError::InvalidData("Received invalid refresh_token expiration".to_string())
-                })? <= REFRESH_WITHIN_SECS;
-
-            match (should_refresh_access, should_generate_new_tokens) {
-                // Generate new tokens if the refresh_token is close to being expired.
-                (_, true) => {
-                    let kp = cluster_info.keypair().clone();
-
-                    let (new_access_token, new_refresh_token) =
-                        generate_auth_tokens(&mut auth_service_client, kp.as_ref()).await?;
-
-                    *access_token.lock().unwrap() = new_access_token.clone();
-                    refresh_token = new_refresh_token;
-
-                    num_full_refreshes += 1;
-                    datapoint_info!(
-                        "auth_tokens_update_loop-tokens_generated",
-                        ("url", url, String),
-                        ("count", num_full_refreshes, i64),
-                    );
-                }
-                // Invoke the refresh_access_token method if the access_token is close to being expired.
-                (true, _) => {
-                    let new_access_token =
-                        refresh_access_token(&mut auth_service_client, refresh_token.clone())
-                            .await?;
-                    *access_token.lock().unwrap() = new_access_token;
-
-                    num_refresh_access_token += 1;
-                    datapoint_info!(
-                        "auth_tokens_update_loop-refresh_access_token",
-                        ("url", url, String),
-                        ("count", num_refresh_access_token, i64),
-                    );
-                }
-                // Sleep and do nothing if neither token is close to expired,
-                (false, false) => sleep(sleep_interval).await,
-            }
-        }
-
-        Ok(())
+/// Tries to refresh the access token or run full-reauth if needed.
+/// This method writes to access_token if refresh is done.
+/// It overwrites the refresh token if full-reauth is run.
+pub async fn maybe_refresh_auth_tokens(
+    auth_service_client: &mut AuthServiceClient<Channel>,
+    access_token: &Arc<Mutex<Token>>,
+    refresh_token: Token,
+    cluster_info: &Arc<ClusterInfo>,
+    connection_timeout: &Duration,
+) -> crate::proxy::Result<(
+    Option<Token>, // access token
+    Option<Token>, // refresh token
+)> {
+    const REFRESH_WITHIN_S: u64 = 5;
```

Comment:
nit: should pass this in and should be related to how often refresh happens

Reply:
for instance, if we change refresh period to 1 minute, there could be a period of time where the refresh token is bad.

the caller should be responsible for saying when the refresh should happen within

Reply:
makes sense

Reply:
makes sense